L112-node-server-interceptors.md

```diff
@@ -0,0 +1,244 @@
+
+```
+underlying ServerCall created ->
+  interceptorA call ->
+    interceptorB call ->
+      interceptorC call ->
+        start called ->
+      interceptorC start ->
+    interceptorB start ->
+  interceptorA start ->
+underlying ServerCall start ->
+  interceptorA onReceiveMetadata ->
+    interceptorB onReceiveMetadata ->
+      interceptorC onReceiveMetadata ->
+        method handler invocation ->
+      interceptorC outbound ->
+    interceptorB outbound ->
+  interceptorA outbound ->
+underlying ServerCall ->
+  interceptorA inbound ->
+    interceptorB inbound ->
+      interceptorC inbound ->
+        method handler
+```
+
+## Rationale
+
+### Interception API
+
+The `Responder` and `ServerListener` APIs are modeled very closely on the `Requester` and `Listener` APIs in the [Client Interceptors design][L5], with small modifications to account for the differences in events that are asymmetrical between the client and the server. This should make it easy for developers who are already familiar with the client interceptor API to create server interceptors.
+
+#### The `start` method
+
+The `Responder#start` method is the only one that does not correspond to any actual network activity. At the HTTP/2 level, the beginning of an incoming call is signalled by receiving the request headers for that call. However, this start method provides a simple way to register the listeners in the proper order after all interceptors have been invoked and to allow those listeners to handle the incoming metadata.
+
+#### The `onCancel` method
+
+The `ServerListener#onCancel` method is "called when the request ends for any reason, including when cancelled by the remote end", which does not strictly match the meaning of the term "cancel". The purpose of handling cancellation on the server is to let the handler know when it does not need to do any further processing to send information to the client, because the client will not see it. So, it is useful to notify whenever that condition is true, no matter what triggered it.
+
+### Interceptors only called for registered methods
+
+For consistency with the client interceptors design, interceptors will see un-serialized message objects. The deserialization functions are only known for registered methods, so interceptors can only be called for registered methods. There is space for future APIs to fill that gap without breaking anything.
+
+### Single injection point
+
+The client interceptor design has interceptor lists and interceptor provider lists, and interceptors specified at client construction vs interceptors specified for individual call invocations. In contrast, this design only specifies that an interceptor list can be provided at server construction. The primary purpose of this design is to solidify the interceptor part of the design. Additional injection points for interceptors could be specified in the future, and the way they interact with each other can be decided at that time.
```

Comment:
This is probably one of the areas that is unique to server interceptors - which by design shouldn't support per-call inceptor injection as any inbound operations/events are supposed to be intercepted before any business logic is invoked (as part of the service method implementation).

Reply:
I updated this section to try to clarify it.